### PR TITLE
sort on TableWrite

### LINKED ISSUE #3692: Table.write should always write ordered if the table has a key

### PR DESCRIPTION
**Pull request by patrick-schultz**
Fixes #3692. I also factored out some of the TableKeyBy logic and tweaked it. If a table is already backed by an OrderedRVD with key ("a", "b", "c") and partition key ("a"), and we do a keyBy(["a", "b"], ["a", "b"]), then no work needs to be done, but we want the OrderedRVDType of the underlying OrderedRVD to remember the stronger invariants it satisfies. That way if we later keyBy(["a", "b", "c"], ["a"]), we don't have to do any work.

### CODE REVIEW

File: src/main/scala/is/hail/expr/Relational.scala [hunk 1]
```diff
@@ -1601,6 +1601,38 @@ case class TableValue(typ: TableType, globals: BroadcastRow, rvd: RVD) {
...
+  // key fields. The keys and nPartitionKeys arguments are regarded as
+  // constraints the returned OrderedRVD must satisfy. If rvd is already an
+  // OrderedRVD satisfying stronger constraints, we don't need to do do any
+  // work, but we also don't want to throw away the knowledge of the stronger
+  // invariants.
+  def toOrderedRVD(keys: Array[String], nPartitionKeys: Int): OrderedRVD = {
+    rvd match {
+      case orvd: OrderedRVD
+        if orvd.typ.key.startsWith(keys)
+        && orvd.typ.partitionKey.length <= nPartitionKeys =>
```
Comment (danking) [inline src/main/scala/is/hail/expr/Relational.scala]: shouldn't this be the opposite?

File: src/main/scala/is/hail/rvd/OrderedRVD.scala [hunk 2]
```diff
@@ -99,7 +99,6 @@ class OrderedRVD(
 
     require(ordType.rowType == typ.rowType)
     require(ordType.kType isPrefixOf typ.kType)
-    require(newPartitioner.pkType isIsomorphicTo ordType.pkType)
```
Comment (danking) [inline src/main/scala/is/hail/rvd/OrderedRVD.scala]: why can we remove this requirement?

File: src/main/scala/is/hail/expr/Relational.scala (hunk 1 above)
Comment (patrick-schultz) [reply src/main/scala/is/hail/expr/Relational.scala]: Suppose the table's rvd is ordered with key `((pk1: String), k2: Int, k3: String, k3: Int)`, and we call `toOrderedRVD(Array("pk1", "pk2", "k3"), 2)`. The old range bounds would be something like `[("A", "C"], ("C", "F"], ["H", "I"]]`. That satisfies the request that all keys sharing the same pk1 and k2 will live in the same partition.

**Comment by tpoterba** [comment]
bump

**Comment by patrick-schultz** [comment]
I backed out the more subtle pieces to get a basic sort on write in now. Sorry, I should have done that sooner. When I didn't have an easy answer for why it was safe to remove that requirement, I realized I needed to do some work solidifying the partitioner semantics. I'm getting closer on that front, but I'm constantly amazed how tricky it is!

**Pull request merged**